apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/observer/ShardObserver.java

```diff
@@ -45,33 +47,64 @@ public void onTrimDone(@Observes @Sync TrimData trimData) {
+            return true;
+        }
+        return false;
+    }
+    
+    private boolean isTimeForShard(int lastTrimBlockHeight, int blockchainHeight, HeightConfig currentConfig) {
         int shardingFrequency = 1;
-        if (shardingEnabled && !isShardingOff) {
-            if (!blockchainConfig.isJustUpdated()) {
-                // config didn't change from previous trim scheduling
-                shardingFrequency = currentConfig.getShardingFrequency();
-            } else {
-                // TODO: YL after separating 'shard' and 'trim' logic, we can remove 'isJustUpdated()' usage and checking
-                // config has changed from previous trim scheduling, try to get previous 'shard frequency' value
-                shardingFrequency = blockchainConfig.getPreviousConfig().isPresent()
-                        && blockchainConfig.getPreviousConfig().get().isShardingEnabled() ?
-                        blockchainConfig.getPreviousConfig().get().getShardingFrequency() // previous config
-                        : currentConfig.getShardingFrequency(); // fall back
-            }
-            log.debug("Check shard conditions: ? [{}],  lastTrimBlockHeight = {}, blockchainHeight = {}"
-                            + ", shardingFrequency = {} ({})",
-                    shardingFrequency != 0 ?
-                            lastTrimBlockHeight % shardingFrequency == 0 : "zeroDivision",
-                    lastTrimBlockHeight, blockchainHeight,
-                    shardingFrequency, blockchainConfig.isJustUpdated());
-            if (lastTrimBlockHeight != 0 && lastTrimBlockHeight % shardingFrequency == 0) {
+        if (!blockchainConfig.isJustUpdated()) {
+            // config didn't change from previous trim scheduling
+            shardingFrequency = currentConfig.getShardingFrequency();
+        } else {
+            // TODO: YL after separating 'shard' and 'trim' logic, we can remove 'isJustUpdated()' usage and checking
+            // config has changed from previous trim scheduling, try to get previous 'shard frequency' value
+            shardingFrequency = blockchainConfig.getPreviousConfig().isPresent()
+                    && blockchainConfig.getPreviousConfig().get().isShardingEnabled()
+                    ? blockchainConfig.getPreviousConfig().get().getShardingFrequency() // previous config
+                    : currentConfig.getShardingFrequency(); // fall back
+        }
+        log.debug("Check shard conditions: ? [{}],  lastTrimBlockHeight = {}, blockchainHeight = {}"
+                + ", shardingFrequency = {} ({})",
+                shardingFrequency != 0
+                        ? lastTrimBlockHeight % shardingFrequency == 0 : "zeroDivision",
+                lastTrimBlockHeight, blockchainHeight,
+                shardingFrequency, blockchainConfig.isJustUpdated());
+        
+        //Q. how much blocks we ould be late? 1/4 of frequiency is OK?
```

Comment:
First, we can be late if we miss that trim event and it is quite possible in case of restart. Second, one event only that should trigger all other events is new block acceptance. Triggering sharding from trim and then calling trim from shard is cause of code instability or dead locks. Third, trim is not life-important procedure but we already have too much things bound to it. So we have to prepare for unbinding